### ISSUE TITLE
[HOLD for payment 2023-02-03] [$1000] [Image] The attachment PDF preview is not shown offline reported by @Santhosh-Sellavel 

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12512]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open any chat in the new dot
2. Go offline
3. Click :heavy_plus_sign: in composer -> Add Attachments
4. Select a PDF attachment

## Expected Result:
Should show a preview like we show image preview

## Actual Result:
Failed to load error

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.2.24-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/200342114-636d423b-48cb-4749-b154-cba6b3310446.mov

<img width="960" alt="4" src="https://user-images.githubusercontent.com/43996225/200342156-2472619b-ffff-4c3c-954f-e14f17a047e0.png">

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @Santhosh-Sellavel 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1667782878782179

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0101e6d9faf1232624</li>
        <li>Upwork Job ID: 1600152502185627648</li>
        <li>Last Price Increase: 2022-12-08</li>
    </ul>
</details>

**Comment by mallenexpensify** [comment]
I'm heading OOO, reassigning.

**Comment by davidcardoza** [comment]
I was able to reproduce this. Assigning an engineer to triage - can this be worked on by an external contributor?

**Comment by joelbettner** [comment]
> can this be worked on by an external contributor

Yes, I believe this can be worked on by an external contributor.

**Comment by JmillsExpensify** [comment]
cc @youssef-lr I am kind of thinking that we need to have someone holistically approach all these attachment uploads cases. What do you think about us creating one central issue for all the remaining bugs and improvements we need to make? I'm happy to get it started.

**Comment by sketchydroide** [comment]
@JmillsExpensify @youssef-lr should this still be an external?

**Comment by juzoace** [comment]
@JmillsExpensify is this still up for grabs?

**Comment by davidcardoza** [comment]
I believe this should be external, but I will wait for @JmillsExpensify to confirm.

**Comment by youssef-lr** [comment]
> cc @youssef-lr I am kind of thinking that we need to have someone holistically approach all these attachment uploads cases. What do you think about us creating one central issue for all the remaining bugs and improvements we need to make? I'm happy to get it started.

Sounds good to me, let's do it!

I do believe this can be `External` as well.

**Comment by JmillsExpensify** [comment]
@davidcardoza Yeah external is fine!

**Comment by davidcardoza** [comment]
Job posted - https://www.upwork.com/jobs/~01182973b76b5df6f6

**Comment by sketchydroide** [comment]
not sure if we should update the price?

**Comment by puneetlath** [comment]
@sketchydroide this is one of the oldest issues in the /App repo. To help us clear out the large backlog of bugs, can you:  
- Decide whether any proposals currently meet our guidelines and can be approved as-is 
- For any that can't, please take this issue internal and treat it as one of your highest priorities 
- If you have any questions, don't hesitate to start a discussion in #bug-zero

**Comment by juzoace** [comment]
- With what I can see, there's an error indicated in the network tab as 51103eacf3bef7f35f0812db4ea611f7.js. I'll still dig deeper into it. 

 I'm guessing the current design might need to be changed because of the information I saw here:  https://github.com/wojtekmaj/react-pdf/issues/1061

I'm guessing there might not be a way to show PDF when offline for now. I tried using a worker but instead the pdf also failed to display with the network. You can still keep it internal for now while I keep an eye for a response from the contributors of react-pdf

**Comment by juzoace** [comment]
Okay @rushatgabhane, I'll keep you posted.

**Comment by thomas-coldwell** [comment]
Hi I'm from Margelo - I'll be able to take a look at this issue today 🙂

**Comment by davidcardoza** [comment]
Great, thank you @thomas-coldwell

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by dbond09** [comment]
### Proposal

The reason this issue occurs is that `react-pdf` uses a web worker to render the preview. The constructor of said web worker takes an URL as its argument which points to the code that will be executed by the worker. The URL points to a `.js` file on the server, which was not bundled with Webpack as it should have been. So when we try to render the PDF preview in offline mode, the web worker tries to fetch the file from the server which is inaccessbile.

My proposed solution is as follows. First, we force Webpack to bundle the web worker code by importing it in `PDFView/index.js` with the raw-loader. Then, as we now have the code in memory, we create a `Blob` out of it, to which we assign a URL. The last step is to set the new URL to be used by the web worker, which is easily done by setting the `workerSrc` global variable in `react-pdf`. As the web worker now doesn't have to send a request to the server, the preview will be working in offline mode too.

Here you can see this solution implemented:

https://user-images.githubusercontent.com/22153853/203656923-319c5ea5-4ad3-4422-bdea-a69ac66d7f4d.mov

**Comment by sketchydroide** [comment]
> @sketchydroide this is one of the oldest issues in the /App repo. To help us clear out the large backlog of bugs, can you:
> 
> * Decide whether any proposals currently meet our guidelines and can be approved as-is
> * For any that can't, please take this issue internal and treat it as one of your highest priorities
> * If you have any questions, don't hesitate to start a discussion in #bug-zero

@puneetlath not sure if this is correct, as this GH is not even 3 weeks yet, not sure how it's one of the oldest issues 🤔 
That said.
We just started getting some solutions coming, and they seem to be from very comptent sources, including Margelo. I think we need to give a bit more time to see how this develop, it seems to me that we will have some good choices in a couple of days.

**Comment by sketchydroide** [comment]
That does look good @dbond09, like @rushatgabhane show us the code you used to make that happen :)

**Comment by dbond09** [comment]
@rushatgabhane This is not applicable for native, as that uses a different implementation in `PDFView/index.native.js`, which doesn't use web workers, so the issue doesn't seem to be occurring there.

@sketchydroide Updated the proposal with the code.

**Comment by sketchydroide** [comment]
any update @rushatgabhane :)

**Comment by sketchydroide** [comment]
Friendly bump @rushatgabhane

**Comment by sketchydroide** [comment]
no worries, thanks for the honesty, I'll reasign this

**Comment by sketchydroide** [comment]
Melvin really like you @rushatgabhane

**Comment by sketchydroide** [comment]
@aimane-chnaif thanks for joining can you please check @dbond09 proposal [here](https://github.com/Expensify/App/issues/12512#issuecomment-1325733007)

**Comment by JmillsExpensify** [comment]
I wanted to check back on @dbond09's proposal. Have we been able to review it yet?

**Comment by sketchydroide** [comment]
Yep, let us know when you think you;ll be able to check on this @aimane-chnaif

**Comment by sketchydroide** [comment]
Thanks

**Comment by MitchExpensify** [comment]
@sketchydroide  this is one of the oldest issues in the /App repo. To help us clear out the large backlog of bugs, can you:

- Decide whether any proposals currently meet our guidelines and can be approved as-is
- For any that can't, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #bug-zero

**Comment by MitchExpensify** [comment]
I've updated the [Upwork Job](https://www.upwork.com/jobs/~01182973b76b5df6f6) post and issue to reflect $1000

**Comment by dbond09** [comment]
> @dbond09 your solution looks promising but only for web. Does it work on mWeb as well? Also, do you have proposal for native platforms? On native, it shows infinite loading when offline. We need to make it consistent throughout all platforms.

@aimane-chnaif I can confirm that the proposal fixes the Web, mWeb and Desktop platforms. For iOS and Android, this issue seems to have been also resolved.

**Comment by marcaaron** [comment]
Is there some alternative we can use to this `raw-loader!` syntax? That is pretty non-standard for our app. It feels like there should be another way to get the worker code bundled.

**Comment by marcaaron** [comment]
Poked around a bit and something like this worked for me:

```diff
diff --git a/config/webpack/webpack.common.js b/config/webpack/webpack.common.js
index ef1f46e838..6f2cd01e6a 100644
--- a/config/webpack/webpack.common.js
+++ b/config/webpack/webpack.common.js
@@ -141,6 +141,10 @@ const webpackConfig = ({envFile = '.env', platform = 'web'}) => ({
                     new RegExp(`node_modules/(?!(${includeModules})/).*|.native.js$`),
                 ],
             },
+            {
+                test: new RegExp('node_modules/pdfjs-dist/legacy/build/pdf.worker.js'),
+                type: 'asset/source',
+            },

             // Rule for react-native-web-webview
             {
```

Which then enabled importing as a string here:

```diff
diff --git a/src/components/PDFView/index.js b/src/components/PDFView/index.js
index 485b40c1e9..f44baa00c5 100644
--- a/src/components/PDFView/index.js
+++ b/src/components/PDFView/index.js
@@ -1,7 +1,8 @@
 import _ from 'underscore';
 import React, {Component} from 'react';
 import {View, Dimensions} from 'react-native';
-import {Document, Page} from 'react-pdf/dist/esm/entry.webpack';
+import {Document, Page, pdfjs} from 'react-pdf/dist/esm/entry.webpack';
+import pdfWorkerSource from 'pdfjs-dist/legacy/build/pdf.worker';
 import FullScreenLoadingIndicator from '../FullscreenLoadingIndicator';
 import styles from '../../styles/styles';
 import variables from '../../styles/variables';
```

**Comment by marcaaron** [comment]
>For iOS and Android, this issue seems to have been also resolved.

iOS and Android code is not being touched by current proposal. AFAICT we are not experiencing any issue there. I just tested on staging iOS native and was able to view the PDF preview fine with no error message while offline.

>Also, do you have proposal for native platforms? On native, it shows infinite loading when offline. We need to make it consistent throughout all platforms.

Not sure what is meant by this it looks totally consistent to me. It could be better. I think there are some open issues to improve the "Uploading Attachment..." text that happens, but it's the same on web.

![2022-12-12_14-26-30](https://user-images.githubusercontent.com/32969087/207196620-f0cf52e3-5401-44d7-8732-587d66e7a35e.png)

**Comment by marcaaron** [comment]
Tested on mWeb and the solution works there as well. Since I already have a draft PR going I'm just going to grab this one. I think @dbond09 has the right idea and should be compensated.

**Comment by marcaaron** [comment]
@aimane-chnaif idk what you are doing there but it doesn't look like the repro steps for this issue.

**Comment by trjExpensify** [comment]
Awesome, thanks for that @marcaaron. PR is in review, so adding the `Reviewing` label. 👍

**Comment by marcaaron** [comment]
Just summarizing the status of this. We had to revert the linked PR because of a content security policy warning that was only detectable on staging. Going to bring up a conversation about how to properly get around that issue and then re-raise the PR after.

**Comment by marcaaron** [comment]
Hmm I think the comments above and the title are incorrect as the PR was reverted and definitely did not make it onto production - I guess I will edit and hide them? This is the PR that was merged to production -> https://github.com/Expensify/App/pull/13707

**Comment by roryabraham** [comment]
Solution was reverted IIRC

**Comment by marcaaron** [comment]
Ah, sorry I dropped the ball here. We needed to ask about the CSP stuff in #infra will do that now.

**Comment by marcaaron** [comment]
Summarized here https://expensify.slack.com/archives/C03TQ48KC/p1673388443464709 thanks for the reminder @roryabraham

**Comment by trjExpensify** [comment]
Wagwan! Can we put this back on `Daily` as it's a 4+ week old `Bug` breaking WAQ?

**Comment by JmillsExpensify** [comment]
This is an interesting one. We should have been highlighting this a month ago when the `awaiting payment` and `reviewing` labels came off. I'm not really clear how this issue evaded our WAQ filters, does anyone else?

**Comment by mallenexpensify** [comment]
The `Reviewing` label was only removed six days ago so today is the first time we're seeing for the weekly resync (which I'm working on now, hence this comment)
https://github.com/Expensify/App/issues/12512#event-8195228819

**Comment by marcaaron** [comment]
@JmillsExpensify maybe because it was reopened?

**Update:** I have been waiting on the resolution to this thread [here](https://expensify.slack.com/archives/C03TQ48KC/p1673388443464709). I think we have agreed that it is safe to change the CSP so I'll try to raise a PR for that today. @roryabraham gave me some tips on Cloudflare workers so I'll add it to my list of todos.

**Comment by JmillsExpensify** [comment]
Yeah, that makes sense. Thanks for the urgency on this one regardless!

**Comment by marcaaron** [comment]
CSP changes have been deployed. Re-opening the original fix PR [here](https://github.com/Expensify/App/pull/14561).

**Comment by marcaaron** [comment]
I don't think this was a regression was it? There's no real investigation to do. This has just never worked correctly offline.

**Comment by marcaaron** [comment]
Checked off the first 3 boxes. We should create a test rail step for this though.

**Comment by marcaaron** [comment]
Think this one is just waiting for payment + regression tests.

**Comment by davidcardoza** [comment]
payment sent to @aimane-chnaif. @marcaaron is a regression test necessary per your comment here 

>I don't think this was a regression was it? There's no real investigation to do. This has just never worked correctly offline.

**Comment by marcaaron** [comment]
Yes, we should create a regression test even though this was not _specifically caused by merging another PR_. 

If I am understanding the BZ checklist instructions [here](https://stackoverflowteams.com/c/expensify/questions/14574) we should. In this case, there was just no specific PR that caused and regression or additional conversation to be had.

I also understand that [we changed this recently so it should be responsibility of whoever fixes the issue](https://expensify.slack.com/archives/C01GTK53T8Q/p1674815633050099) (which was me 😅) but then BZ is supposed to create the issue to add it to test rail?

Anyway, the steps would be:

1. Open any chat in New Expensify on Web
1. Go offline
1. Click ➕ in composer -> Add Attachments
1. Select a PDF attachment
1. Verify a preview of the PDF can be seen

**Comment by davidcardoza** [comment]
Ya, admittedly, I have yet to run a regression test, it's a bit unclear to me when a regression test should take place or not. Getting a buddy check here - https://expensify.slack.com/archives/C01SKUP7QR0/p1675814496893819

**Comment by davidcardoza** [comment]
A regression test was added as the previous test was pretty similar to the offline PDF case in this test case:
https://expensify.testrail.io/index.php?/cases/view/1971194&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229062

Updated to:

**Offline**
1. Open a conversation
2. Go offline
3. Click on the plus sign in composer -> Add Attachments
4. Select a PDF attachment
5. Verify a preview of the PDF can be seen

**Comment by Santhosh-Sellavel** [comment]
@davidcardoza reporting bonus is missed out here!

**Comment by davidcardoza** [comment]
@Santhosh-Sellavel offer sent.

**Comment by Santhosh-Sellavel** [comment]
Thanks @davidcardoza Accepted!

**Comment by kbecciv** [comment]
Issue is reproductible in build 1.2.87.0

![Screenshot_2023-03-17-20-06-08-44_4f9154176b47c00da84e32064abf1c48](https://user-images.githubusercontent.com/93399543/225992665-0205792e-6606-4908-9bc7-0d399498250d.jpg)

**Comment by kbecciv** [comment]
Yes, you are correct, I'm trying  to preview PDF which was already sent as attachment.

**Comment by trjExpensify** [comment]
Doesn't this one come down to caching on web/desktop, which is being worked on elsewhere?

**Comment by trjExpensify** [comment]
Right, but presumably the work we do to support image caching is going to be extended to other file types? I think @thomas-coldwell has it on his radar as a follow-up.

**Comment by kbecciv** [comment]
@marcaaron Logged separate issue here  https://github.com/Expensify/App/issues/16694
